File: dydwo0740/baekjoon/18430.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean flag = false;
    static int ans = 0;
    static int N;
    static int M;

    static int[][] col = {{-1, 0}, {-1, 0}, {0, 1}, {0, 1}};
    static int[][] row = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        int[][] board = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        dfs(board, new boolean[N][M], 0, 0);

        bw.write(ans + "\n");
        bw.flush();
    }

    public static void dfs(int[][] board, boolean[][] visited, int x, int sum) {
        ans = Math.max(ans, sum);



        for (int i = x; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (!visited[i][j]) {
                    for (int s = 0; s < 4; s++) {
                        int nx1 = i + row[s][0];
                        int ny1 = j + col[s][0];
                        int nx2 = i + row[s][1];
                        int ny2 = j + col[s][1];

                        if (0 <= nx1 && nx1 < N && 0 <= ny1 && ny1 < M) {
                            if (0 <= nx2 && nx2 < N && 0 <= ny2 && ny2 < M) {
                                if (!visited[nx1][ny1] && !visited[nx2][ny2]) {
                                    visited[i][j] = true;
                                    visited[nx1][ny1] = true;
                                    visited[nx2][ny2] = true;
                                    //System.out.println("s + \" \" + i + \" \"  + j = " + s + " " + i + " " + j + " " + (sum + board[i][j] * 2 + board[nx1][ny1] + board[nx2][ny2]));
                                    dfs(board, visited, i, sum + board[i][j] * 2 + board[nx1][ny1] + board[nx2][ny2]);
                                    visited[i][j] = false;
                                    visited[nx1][ny1] = false;
                                    visited[nx2][ny2] = false;
                                }
                            }
                        }
                    }
                }
            }
        }
    }


}

File: qsunki/baekjoon/18430.py
import sys


def back(x, y, sum_):
    if y == m:
        x += 1
        y = 0
    if x == n:
        global ans
        ans = max(ans, sum_)
        return
    if not visited[x][y]:
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if visited[x][ny] or visited[nx][y]:
                continue
            visited[x][y] = True
            visited[x][ny] = True
            visited[nx][y] = True
            back(x, y + 1, sum_ + graph[x][y] * 2 + graph[x][ny] + graph[nx][y])
            visited[x][y] = False
            visited[x][ny] = False
            visited[nx][y] = False
    back(x, y + 1, sum_)


n, m = map(int, sys.stdin.readline().split())
graph = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
visited = [[False] * m for _ in range(n)]
dx = [1, 1, -1, -1]
dy = [1, -1, 1, -1]
ans = 0
back(0, 0, 0)
print(ans)
